eventdata/parameter_sources/interval_query_source.py

```diff
@@ -0,0 +1,136 @@
+            self._cache = params['cache']
+
+        if 'index_pattern' in params.keys():
+            self._index_pattern = params['index_pattern']
+
+        if 'type' in params.keys():
+            self._type = params['type']
+        
+        if 'fieldname' in params.keys():
+            self._fieldname = params['fieldname']
+
+        if 'body' in params.keys():
+            self._body = params['body']
+
+            # Verify that body contains `query.bool.must` key.
+            if not params['body']['query']['bool'] or ('must' not in params['body']['query']['bool'].keys()):
+                raise ConfigurationError("Parameter 'body' must contain `query.bool.must` key.");
+        else:
+            raise ConfigurationError("Parameter 'body' must be specified.");
+
+        if 'min_interval_size_pct' in params.keys():
+            self._min_interval_size_pct = float(params['min_interval_size_pct']);
+            if self._min_interval_size_pct <= 0:
+                raise ConfigurationError("Parameter 'min_interval_size_pct' must be > 0.");
+            if self._min_interval_size_pct >= 100:
+                raise ConfigurationError("Parameter 'min_interval_size_pct' must be < 100.");
+        else:
+            raise ConfigurationError("Parameter 'min_interval_size_pct' must be specified.");
+
+        if 'max_interval_size_pct' in params.keys():
+            self._max_interval_size_pct = float(params['max_interval_size_pct']);
+            if self._max_interval_size_pct <= self._min_interval_size_pct:
+                raise ConfigurationError("Parameter 'max_interval_size_pct' must be > 'min_interval_size_pct'.");
+            if self._max_interval_size_pct > 100:
+                raise ConfigurationError("Parameter 'max_interval_size_pct' must be <= 100.");
+        else:
+            raise ConfigurationError("Parameter 'max_interval_size_pct' must be specified.");
+
+    def partition(self, partition_index, total_partitions):
+        return self
+
+    def size(self):
+        return 1
+
+    def params(self):
+        key = "{}_{}".format(self._index_pattern, self._fieldname);
+        if key in fs.global_fieldstats.keys():
+            stats = fs.global_fieldstats[key];
+        else:
+            raise ParameterSourceError("No statistics found for field `{}` in index pattern `{}`.".format(self._fieldname, self._index_pattern));
+
+        range_max = stats['max'];
+        range_min_upper = int(stats['max'] - ((stats['max'] - stats['min']) * self._min_interval_size_pct / 100.0));
+        delta = (stats['max'] - stats['min']) * ((self._max_interval_size_pct - self._min_interval_size_pct) / 100.0);
+        range_min = range_min_upper - int(delta * random.random());
+        body = copy.deepcopy(self._body);
+
+        # Rewrite query to include new range filter based on range_min and range_max.
+        # This assumes that the body contains `query.bool.must` key.
+        if not isinstance(body['query']['bool']['must'], list):
```

Comment:
Move this to the constructor?

Reply:
Will keep it where it is, at least for the time being.